packages/apollo-server-core/src/plugin/usageReporting/__tests__/plugin.test.ts

```diff
@@ -203,8 +243,129 @@ describe('end-to-end', () => {
+          },
+        },
+      });
+      expect(context.metrics.includeOperationInUsageReporting).toBe(true);
+      expect(context.metrics.fieldLevelInstrumentation).toBe(true);
+      expect(Object.keys(report!.tracesPerQuery)).toHaveLength(1);
+      expect(
+        containsFieldExecutionData(Object.values(report!.tracesPerQuery)[0]!),
+      ).toBe(true);
+    });
+
+    it('exclude based on operation name', async () => {
+      const { report, context } = await runTest({
+        pluginOptions: {
+          fieldLevelInstrumentation: async (
+            requestContext: GraphQLRequestContextDidResolveOperation<any>,
+          ) => {
+            await new Promise<void>((res) => setTimeout(() => res(), 1));
+            return requestContext.request.operationName === 'not_q';
+          },
+        },
+        schemaShouldBeInstrumented: false,
+      });
+      // We do get a report about this operation; we just don't have field
+      // execution data (as trace or as TypeStat).
+      expect(context.metrics.includeOperationInUsageReporting).toBe(true);
+      expect(context.metrics.fieldLevelInstrumentation).toBe(false);
+      expect(Object.keys(report!.tracesPerQuery)).toHaveLength(1);
+      expect(
+        containsFieldExecutionData(Object.values(report!.tracesPerQuery)[0]!),
+      ).toBe(false);
+    });
+
+    it('pass a number', async () => {
+      const total = 100;
+      const fraction = 0.35;
+      let actualFieldLevelInstrumentation = 0;
+      let actualContainsFieldExecutionData = 0;
+
+      for (let i = 0; i < total; ++i) {
+        const { report, context } = await runTest({
+          pluginOptions: {
+            fieldLevelInstrumentation: fraction,
+          },
+          schemaShouldBeInstrumented: null,
+        });
+        // We do get a report about this operation; we just don't have field
+        // execution data (as trace or as TypeStat).
+        expect(context.metrics.includeOperationInUsageReporting).toBe(true);
+        if (context.metrics.fieldLevelInstrumentation === true) {
+          actualFieldLevelInstrumentation++;
+        }
+        expect(Object.keys(report!.tracesPerQuery)).toHaveLength(1);
+        if (
+          containsFieldExecutionData(Object.values(report!.tracesPerQuery)[0]!)
+        ) {
+          actualContainsFieldExecutionData++;
+        }
+      }
+
+      // Make sure the number of reports that contain field execution data
+      // matches the number of times that the plugin set the
+      // fieldLevelInstrumentation metrics flag.
+      expect(actualContainsFieldExecutionData).toBe(
+        actualFieldLevelInstrumentation,
+      );
+      const expected = fraction * total;
+      // If it strays from the expected amount of 35 by too far we fail.
+      expect(actualFieldLevelInstrumentation).toBeGreaterThanOrEqual(
+        0.6 * expected,
+      );
+      expect(actualFieldLevelInstrumentation).toBeLessThanOrEqual(
+        1.4 * expected,
+      );
```

Comment:
Smells like a flaky test, can't we mock `Math.random()` to get deterministic results in our tests?

Reply:
Interesting point! https://stackoverflow.com/questions/41570273/how-to-test-a-function-that-output-is-random-using-jest or https://www.npmjs.com/package/jest-mock-random might do the trick here.

Might also remove the test — going to be changing how this is implemented anyway.

Reply:
Mocked Math.random.